crates/sui-graphql-rpc/src/types/checkpoint.rs

```diff
@@ -402,3 +451,107 @@ impl Checkpointed for Cursor {
         self.checkpoint_viewed_at
     }
 }
+
+#[async_trait::async_trait]
+impl Loader<SeqNumKey> for Db {
+    type Value = Checkpoint;
+    type Error = Error;
+
+    async fn load(&self, keys: &[SeqNumKey]) -> Result<HashMap<SeqNumKey, Checkpoint>, Error> {
+        use checkpoints::dsl;
+
+        let checkpoint_ids: BTreeSet<_> = keys
+            .iter()
+            .filter_map(|key| {
+                if let Some(viewed_at) = key.checkpoint_viewed_at {
+                    // Filter out keys querying for checkpoints after their own consistency cursor.
+                    (viewed_at >= key.sequence_number).then_some(key.sequence_number as i64)
+                } else {
+                    Some(key.sequence_number as i64)
+                }
+            })
+            .collect();
+
+        let checkpoints: Vec<StoredCheckpoint> = self
+            .execute_repeatable(move |conn| {
+                conn.results(move || {
+                    dsl::checkpoints
+                        .filter(dsl::sequence_number.eq_any(checkpoint_ids.iter().cloned()))
+                })
+            })
+            .await
+            .map_err(|e| Error::Internal(format!("Failed to fetch checkpoints: {e}")))?;
+
+        let checkpoint_id_to_stored: BTreeMap<_, _> = checkpoints
+            .into_iter()
+            .map(|stored| (stored.sequence_number as u64, stored))
+            .collect();
+
+        Ok(keys
+            .iter()
+            .filter_map(|key| {
+                let stored = checkpoint_id_to_stored.get(&key.sequence_number).cloned()?;
+                let checkpoint = Checkpoint {
+                    stored,
+                    checkpoint_viewed_at: key.checkpoint_viewed_at,
+                };
+
+                let digest = &checkpoint.stored.checkpoint_digest;
+                if matches!(key.digest, Some(d) if d.as_slice() != digest) {
+                    None
+                } else {
+                    Some((*key, checkpoint))
+                }
+            })
+            .collect())
+    }
+}
+
+#[async_trait::async_trait]
+impl Loader<DigestKey> for Db {
+    type Value = Checkpoint;
+    type Error = Error;
+
+    async fn load(&self, keys: &[DigestKey]) -> Result<HashMap<DigestKey, Checkpoint>, Error> {
+        use checkpoints::dsl;
+
+        let digests: BTreeSet<_> = keys.iter().map(|key| key.digest.to_vec()).collect();
+
+        let checkpoints: Vec<StoredCheckpoint> = self
+            .execute_repeatable(move |conn| {
+                conn.results(move || {
+                    dsl::checkpoints.filter(dsl::checkpoint_digest.eq_any(digests.iter().cloned()))
```

Comment:
@gegaowp, I wanted to confirm that there weren't plans to partition the checkpoints table? If there are, in order for this to continue working well, we would need to have a way of translating digests to checkpoint sequence numbers without querying every partition.